direct_connect.go

```diff
@@ -0,0 +1,187 @@
+package main
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+	"sync"
+	"time"
+
+	"io"
+	"log"
+	"net/http"
+
+	"golang.zx2c4.com/wireguard/conn"
+	"golang.zx2c4.com/wireguard/device"
+	"golang.zx2c4.com/wireguard/tun/netstack"
+	gtcp "gvisor.dev/gvisor/pkg/tcpip/transport/tcp"
+	gudp "gvisor.dev/gvisor/pkg/tcpip/transport/udp"
+
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+
+	"wiretap/peer"
+	"wiretap/transport/icmp"
+	"wiretap/transport/tcp"
+	"wiretap/transport/udp"
+
+	"github.com/rainforestapp/rainforest-cli/rainforest"
+	"github.com/urfave/cli"
+)
+
+func launchDirectConnect(c cliContext, rfApi *rainforest.Client) error {
+	tunnelId := c.String("tunnel-id")
+	if tunnelId == "" {
+		tunnelId = "default"
+	}
+
+	log.Println("Starting Direct Connect Tunnel for tunnel ID:", tunnelId)
+
+	// Generate a wireguard public/private keypair
+	privateKey, err := wgtypes.GeneratePrivateKey()
+	if err != nil {
+		return cli.NewExitError(fmt.Errorf("failed to generate private key: %w", err), 1)
+	}
+	publicKey := privateKey.PublicKey()
+
+	// Ask the rainforest api to set up our connection, returning the server details
+	serverDetails, err := rfApi.ConfigureDirectConnect(tunnelId, publicKey.String())
+	if err != nil {
+		return cli.NewExitError(fmt.Errorf("failed to configure direct connect: %w", err), 1)
+	}
+
+	var (
+		wg   sync.WaitGroup
+		lock sync.Mutex
+	)
+
+	// Static configuration settings
+	listenPort := 51820
+	persistentKeepaliveInterval := 25
+	mtu := 1420
+	allowedIPs := []string{"198.18.18.1/32", "fd:16::1/128"}
```

Comment:
Because this is purely a userspace implementation of wireguard, I think there is no conflict if the host has the same IPs in use. It just means you won't be able to access the host directly over the tunnel, because it'll be shadowed by the wireguard userspace tunnel. I think this isn't much of a concern.

And yes, that was more or less my logic re: IPv6; I figured there wasn't any harm leaving it here in case we support it in the future it's one less place to change.